### PR TITLE
IGNITE-22183 Recording of Index query events added

### PR DESCRIPTION
**Pull request by oleg-vlsk**
Thank you for submitting the pull request to the Apache Ignite.

In order to streamline the review of the contribution 
we ask you to ensure the following steps have been taken:

### The Contribution Checklist
- [ ] There is a single JIRA ticket related to the pull request. 
- [ ] The web-link to the pull request is attached to the JIRA ticket.
- [ ] The JIRA ticket has the _Patch Available_ state.
- [ ] The pull request body describes changes that have been made. 
The description explains _WHAT_ and _WHY_ was made instead of _HOW_.
- [ ] The pull request title is treated as the final commit message. 
The following pattern must be used: `IGNITE-XXXX Change summary` where `XXXX` - number of JIRA issue.
- [ ] A reviewer has been mentioned through the JIRA comments 
(see [the Maintainers list](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute#HowtoContribute-ReviewProcessandMaintainers)) 
- [ ] The pull request has been checked by the Teamcity Bot and 
the `green visa` attached to the JIRA ticket (see [TC.Bot: Check PR](https://mtcga.gridgain.com/prs.html))

### Notes
- [How to Contribute](https://cwiki.apache.org/confluence/display/IGNITE/How+to+Contribute)
- [Coding abbreviation rules](https://cwiki.apache.org/confluence/display/IGNITE/Abbreviation+Rules)
- [Coding Guidelines](https://cwiki.apache.org/confluence/display/IGNITE/Coding+Guidelines)
- [Apache Ignite Teamcity Bot](https://cwiki.apache.org/confluence/display/IGNITE/Apache+Ignite+Teamcity+Bot)

If you need any help, please email dev@ignite.apache.org or ask anу advice on http://asf.slack.com _#ignite_ channel.

### CODE REVIEW

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 1]
```diff
@@ -1203,7 +1219,7 @@ protected void runQuery(GridCacheQueryInfo qryInfo) {
                         break;
                     }
 
-                    if (type == SCAN || type == INDEX)
+                    if (type == SCAN)
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1222]: Why this change?

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java [hunk 2]
```diff
@@ -1203,7 +1219,7 @@ protected void runQuery(GridCacheQueryInfo qryInfo) {
                         break;
                     }
 
-                    if (type == SCAN || type == INDEX)
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java:1206]: This change might affect IndexQuery performance due to useless generics handling. Let's check how ScanQuery record this event.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteCacheAbstractQuerySelfTest.java [hunk 3]
```diff
@@ -1883,6 +1888,114 @@ public void testFieldsQueryEvents() throws Exception {
...
+                qry.setLocal(true);
+
+            QueryCursor<Cache.Entry<Integer, Type2>> cursor = cache.query(qry);
+
+            cursor.getAll();
+
+            assert readLatch.await(1000, MILLISECONDS);
+            assert execLatch.await(1000, MILLISECONDS);
+
+            if (!evtsDisabled) {
```
Comment (timoninmaxim) [inline modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteCacheAbstractQuerySelfTest.java:1984]: Let's check that qryResults are empty case when `evtsDisabled = true`.

File: modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java (hunk 2 above)
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/processors/cache/query/GridCacheQueryManager.java]: Recording of Index query object reading events moved to `IndexQueryProcessor#queryLocal`.

File: modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteCacheAbstractQuerySelfTest.java (hunk 3 above)
Comment (oleg-vlsk) [reply modules/indexing/src/test/java/org/apache/ignite/internal/processors/cache/IgniteCacheAbstractQuerySelfTest.java:1984]: Condition added.

File: modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java [hunk 4]
```diff
@@ -125,6 +131,38 @@ public <K, V> IndexQueryResult<K, V> queryLocal(
...
 
+                if (cctx.events().isRecordable(EVT_CACHE_QUERY_OBJECT_READ) &&
+                    cctx.gridEvents().hasListener(EVT_CACHE_QUERY_OBJECT_READ)) {
+
+                    final K key = row.getKey();
+                    final V val = row.getValue();
+
+                    CacheObjectContext objCtx = cctx.cacheObjectContext();
+
+                    K key0 = (K)CacheObjectUtils.unwrapBinaryIfNeeded(objCtx, key, keepBinary, false, null);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java:142]: This code duplicates code in the `onHasNext` method.

File: modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java [hunk 5]
```diff
@@ -125,6 +131,38 @@ public <K, V> IndexQueryResult<K, V> queryLocal(
 
                 IgniteBiTuple<K, V> row = currVal;
 
+                if (cctx.events().isRecordable(EVT_CACHE_QUERY_OBJECT_READ) &&
+                    cctx.gridEvents().hasListener(EVT_CACHE_QUERY_OBJECT_READ)) {
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java:135]: Actually, there is no need in checking listeners. In some places this is checked, in some - not. Let's skip this check.

File: modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java [hunk 6]
```diff
@@ -125,6 +131,38 @@ public <K, V> IndexQueryResult<K, V> queryLocal(
 
                 IgniteBiTuple<K, V> row = currVal;
 
+                if (cctx.events().isRecordable(EVT_CACHE_QUERY_OBJECT_READ) &&
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java:134]: Let's cache the result of `isRecordable` for the iterator.

File: modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java (hunk 4 above)
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java]: Made this section of code more concise by using the `unwrap()` method.

File: modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java (hunk 5 above)
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java]: Check for available listeners removed.

File: modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java (hunk 6 above)
Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java]: Boolean variable `isRecordable` is introduced in the `queryLocal()` method.

File: modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java [hunk 7]
```diff
@@ -125,14 +132,39 @@ public <K, V> IndexQueryResult<K, V> queryLocal(
 
                 IgniteBiTuple<K, V> row = currVal;
 
+                if (isRecordable) {
+                    K k = unwrap(row.getKey(), keepBinary);
```
Comment (timoninmaxim) [inline modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java:136]: Code duplication, you already have the K and V in `onHasNext`. Just move events to this method. Note, that filtered entries should not raise the event

Comment (oleg-vlsk) [reply modules/core/src/main/java/org/apache/ignite/internal/cache/query/index/IndexQueryProcessor.java]: Events moved to `onHasNext`.
